File: pkg/transport/dial/dial.go
// Package transportdial provides a convenience factory for creating a
// transport.Transport from environment variables.
//
// This is the recommended quick-start path for most applications.
// Advanced users who need TLS, custom connection pools, or other low-level
// NATS options should construct nats.New / inmem.New directly.
package transportdial

import (
	"context"
	"fmt"
	"os"

	"github.com/ModulationAI/openagentio/pkg/transport"
	"github.com/ModulationAI/openagentio/pkg/transport/inmem"
	"github.com/ModulationAI/openagentio/pkg/transport/nats"
)

type options struct {
	natsName string
}

// Option configures Dial.
type Option func(*options)

// WithNATSName sets the NATS connection name.
// Only used when OPENAGENTIO_TRANSPORT=nats (or default).
func WithNATSName(name string) Option {
	return func(o *options) {
		o.natsName = name
	}
}

// Dial creates a Transport from environment variables.
//
// Env vars:
//   - OPENAGENTIO_TRANSPORT: "nats" (default) or "inmem"
//   - NATS_URL: NATS server URL, default "nats://localhost:4222"
//
// Example:
//
//	tp, err := transportdial.Dial(ctx, transportdial.WithNATSName("echo-agent"))
func Dial(ctx context.Context, opts ...Option) (transport.Transport, error) {
	o := options{}
	for _, fn := range opts {
		fn(&o)
	}

	mode := os.Getenv("OPENAGENTIO_TRANSPORT")
	switch mode {
	case "inmem":
		return inmem.New(), nil
	case "", "nats":
		url := os.Getenv("NATS_URL")
		if url == "" {
			url = "nats://localhost:4222"
		}
		natsOpts := []nats.Option{nats.URL(url)}
		if o.natsName != "" {
			natsOpts = append(natsOpts, nats.Name(o.natsName))
		}
		tp := nats.New(natsOpts...)
		if err := tp.Connect(ctx); err != nil {
			return nil, fmt.Errorf("connect to NATS %s: %w", url, err)
		}
		return tp, nil
	default:
		return nil, fmt.Errorf("unsupported OPENAGENTIO_TRANSPORT=%q", mode)
	}
}

File: sdk/python/src/openagentio/transport/dial.py
"""Transport factory keyed off ``OPENAGENTIO_TRANSPORT``.

Mirrors ``pkg/transport/dial/dial.go``. Reads the transport mode and (for NATS)
the URL from environment variables, constructs the matching driver, and
connects it before returning.

Environment variables:

* ``OPENAGENTIO_TRANSPORT`` -- ``"inmem"`` or ``"nats"``. Empty defaults to
  ``"nats"``. Any other value raises :class:`ValueError`.
* ``NATS_URL`` -- NATS server URL (default ``"nats://localhost:4222"``).
"""
from __future__ import annotations

import os
from dataclasses import dataclass
from typing import Callable

from openagentio.transport.base import Transport
from openagentio.transport.inmem import InMemoryDriver
from openagentio.transport.nats import NATSDriver


_DEFAULT_NATS_URL = "nats://localhost:4222"


@dataclass
class _DialOpts:
    nats_name: str = ""


DialOption = Callable[[_DialOpts], None]


def WithNATSName(name: str) -> DialOption:
    """Set the NATS connection name. Ignored for non-NATS transports."""
    def apply(o: _DialOpts) -> None:
        o.nats_name = name
    return apply


async def dial(*options: DialOption) -> Transport:
    """Construct and connect a Transport based on ``OPENAGENTIO_TRANSPORT``."""
    opts = _DialOpts()
    for opt in options:
        opt(opts)

    mode = os.environ.get("OPENAGENTIO_TRANSPORT", "")
    if mode == "inmem":
        driver = InMemoryDriver()
        await driver.connect()
        return driver

    if mode == "" or mode == "nats":
        url = os.environ.get("NATS_URL", "") or _DEFAULT_NATS_URL
        driver = NATSDriver(url=url, name=opts.nats_name)
        try:
            await driver.connect()
        except Exception as e:
            raise RuntimeError(f"connect to NATS {url}: {e}") from e
        return driver

    raise ValueError(f"unsupported OPENAGENTIO_TRANSPORT={mode!r}")
